net: openthread: Init NCP after USB communication is established

lmaciejonczyk opened the pull request:
The device has sent RESET_POWER_UP message before the communication with the host has been established. It could be observed with pyspinel (with debugging enabled) which displayed `Framing error`.
This commit fixes the bug by initializing NCP after the host stated is ready to communicate.
~~For consistency the behavior is also changed for pure serial communication.~~

rlubos commented on samples/net/openthread/coprocessor/src/main.c:
```diff
@@ -12,39 +12,50 @@ LOG_MODULE_REGISTER(ot_br, LOG_LEVEL_DBG);
+				LOG_ERR("Failed to get Data Terminal Ready line state: %d",
+					ret);
+				continue;
+			}
```
Would it make sense to add a small delay (k_sleep) between the iterations? We're practically busy-looping right now.

lmaciejonczyk replied:
I don't think it will change much. Currently any more important things happen with higher priorities and because main is working as preemptive it can be supplanted. Anyways the application is useless until it connects with a host for what we waits here.

lmaciejonczyk replied:
On the other hand small delay here would let logging thread to process... Ok, let's do that.

rlubos commented on samples/net/openthread/coprocessor/src/main.c:
```diff
@@ -12,39 +12,50 @@ LOG_MODULE_REGISTER(ot_br, LOG_LEVEL_DBG);
+		const struct device *uart_dev;
+		uint32_t dtr = 0U;
+
+		ret = usb_enable(NULL);
```
Just wondering here... this looks like a part of the USB-CDC-specific UART module initialization code, perhaps it would make sense to move it inside `otPlatUartEnable()`? It'd be consistent with the vanilla OT NCP app, where it doesn't bother with low-level platform initialization. I believe the NCP initialization part could then be left as it is and the actual app would be even simpler (so less code to duplicate in case NCP module is a part of a larger app). What do you think?

lmaciejonczyk replied:
Frankly speaking the `usb_enable()` used to be called from `otPlatUartEnable()` but it has been changed by this commit: c6f7d87142739036fdae6dbc722f97ec9acef4df. I agree that calling `usb_enable()` from `otPlatUartEnable()` is cleaner solution.

rlubos replied:
Ok, thank for claryfing. So it seems to be USB limitation then, I guess we have to live with that for now...

lmaciejonczyk replied:
I needed to change the fix for `Framing error`. Previous proposal broke UART communication with `wpantund`.
Eventually I placed `usb_enable()` in `otPlatUartEnable()` and added there the fix.  
The issue, which worried me (#27427 and #27071) , concerned only USB UART console implementation which co-processor sample does not use.

rlubos reviewed (commented):
Looks OK, but perhaps we could discuss the approach here.

lmaciejonczyk committed:
net: openthread: Init NCP after USB communication is established

The device has sent RESET_POWER_UP message before the communication
with the host hadn't been established. It could be observed with
pyspinel which displayed `Framing error`.

This commit fixes the bug by initializing NCP after the host stated
is ready to communicate.

This commit reverts initialization the USB stack into function
otPlatUartEnable to be consistent with others OpenThread platforms.
OpenThread co-processor samples are not affected by #27071 as they use
USB for SPINEL communication with host and not for UART console.

Note:
When co-processor communicates by USB CDC ACM and it is hard reset
(what is happening in current Zephyr OpenThread platform)
the connection needs to be properly handled by the host.

For posix platform used together with RCP it was implemented in:
https://github.com/openthread/openthread/pull/6454

and for NCP:
https://github.com/openthread/wpantund/pull/492 .

Signed-off-by: Lukasz Maciejonczyk <lukasz.maciejonczyk@nordicsemi.no>

lmaciejonczyk commented:
CI is not green. The output message is not clear if it is my fault an I should do sth about it or nor. Anyone help?

edmont commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -159,6 +159,39 @@ otError otPlatUartEnable(void)
 					(void *)&ot_uart);
+
+#ifdef CONFIG_OPENTHREAD_COPROCESSOR_SPINEL_ON_UART_ACM
+	{
```
Why the braces here?

lmaciejonczyk replied:
To create functional block which includes defined variables at the beginning of it which helps treat this functionality as a whole without compile/static analyzer warnings.

edmont commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -159,6 +159,39 @@ otError otPlatUartEnable(void)
+		/* Data Set Ready - the NCP SoC is ready to communicate */
+		(void)uart_line_ctrl_set(ot_uart.dev, UART_LINE_CTRL_DSR, 1);
+	}
+#endif
```
```suggestion
#endif /* CONFIG_OPENTHREAD_COPROCESSOR_SPINEL_ON_UART_ACM */
```

lmaciejonczyk commented:
@carlescufi Could you help with the CI problem?

lmaciejonczyk commented:
I rebased the PR few times and always buildkite/zephyr fail, but it seems the problem is in the CI scripts not the PR itself. Can anyone help?

lmaciejonczyk committed:
samples: openthread: Add test configuration for coprocessor sample

Coprocessor sample was lacked of test configuration which makes
buildkite/zephyr to fail. This commit fixes it by providing test
configuration for this sample.

Signed-off-by: Lukasz Maciejonczyk <lukasz.maciejonczyk@nordicsemi.no>

lmaciejonczyk commented:
> I rebased the PR few times and always buildkite/zephyr fail, but it seems the problem is in the CI scripts not the PR itself. Can anyone help?

The source of CI problem was that the changed code was not covered by any test.
